File: test_customization.js
/**
 * Playwright test for image generation customization dialog
 *
 * Tests:
 * 1. Dialog appears when clicking generate buttons
 * 2. LocalStorage persistence works
 * 3. Form validation works
 * 4. Settings are passed to backend correctly
 */

const { chromium } = require('playwright');

async function testCustomizationDialog() {
    console.log('🧪 Starting Playwright tests for customization dialog...\n');

    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext();
    const page = await context.newPage();

    let testsPassed = 0;
    let testsFailed = 0;

    try {
        // Navigate to the library page
        console.log('📍 Test 1: Navigate to library page');
        await page.goto('http://127.0.0.1:8123/');
        await page.waitForLoadState('networkidle');

        const title = await page.title();
        console.log(`   ✓ Page loaded: ${title}`);
        testsPassed++;

        // Check if there are any books
        const bookLinks = await page.$$('a[href*="/read/"]');

        if (bookLinks.length === 0) {
            console.log('   ⚠️  No books found - skipping reader page tests');
            console.log('   ℹ️  To fully test, process an EPUB first with: uv run reader3.py <book.epub>\n');

            console.log('📊 Test Summary:');
            console.log(`   ✓ Passed: ${testsPassed}`);
            console.log(`   ✗ Failed: ${testsFailed}`);
            console.log(`   ⚠️  Skipped: Reader page tests (no books available)`);

            await browser.close();
            return;
        }

        console.log(`   ✓ Found ${bookLinks.length} book(s)\n`);

        // Navigate to first book
        console.log('📍 Test 2: Navigate to book reader');
        await bookLinks[0].click();
        await page.waitForLoadState('networkidle');
        console.log('   ✓ Reader page loaded\n');
        testsPassed++;

        // Test 3: Check if customization dialog exists
        console.log('📍 Test 3: Check customization dialog exists in DOM');
        const dialog = await page.$('#illustration-dialog');
        if (!dialog) {
            throw new Error('Customization dialog not found in DOM');
        }
        const isHidden = await dialog.evaluate(el => el.classList.contains('hidden'));
        if (!isHidden) {
            throw new Error('Dialog should be hidden by default');
        }
        console.log('   ✓ Customization dialog exists and is hidden by default\n');
        testsPassed++;

        // Test 4: Check dialog fields exist
        console.log('📍 Test 4: Check dialog form fields');
        const numImagesInput = await page.$('#num-images');
        const styleInput = await page.$('#image-style');
        if (!numImagesInput || !styleInput) {
            throw new Error('Dialog form fields not found');
        }
        console.log('   ✓ Number of images input found');
        console.log('   ✓ Style input found\n');
        testsPassed++;

        // Test 5: Open dialog via floating action button
        console.log('📍 Test 5: Open dialog via floating action button');

        // Click the floating action button
        const actionButton = await page.$('#actions-menu-btn');
        if (!actionButton) {
            throw new Error('Floating action button not found');
        }
        await actionButton.click();
        await page.waitForTimeout(200); // Wait for dropdown animation

        // Click generate illustrations button
        const generateBtn = await page.$('#generate-illustrations-btn');
        if (!generateBtn) {
            throw new Error('Generate illustrations button not found');
        }
        await generateBtn.click();
        await page.waitForTimeout(200); // Wait for dialog animation

        // Check if dialog is visible
        const isVisible = await page.$eval('#illustration-dialog',
            el => !el.classList.contains('hidden')
        );
        if (!isVisible) {
            throw new Error('Dialog should be visible after clicking generate button');
        }
        console.log('   ✓ Dialog opens when clicking generate button\n');
        testsPassed++;

        // Test 6: Test form inputs
        console.log('📍 Test 6: Test form inputs');
        await page.fill('#num-images', '5');
        await page.fill('#image-style', 'anime');

        const numValue = await page.inputValue('#num-images');
        const styleValue = await page.inputValue('#image-style');

        if (numValue !== '5' || styleValue !== 'anime') {
            throw new Error('Form inputs not working correctly');
        }
        console.log('   ✓ Number of images set to 5');
        console.log('   ✓ Style set to "anime"\n');
        testsPassed++;

        // Test 7: Test localStorage persistence
        console.log('📍 Test 7: Test localStorage persistence');

        // Confirm dialog (this should save to localStorage)
        const confirmBtn = await page.$('button:has-text("Generate")');

        // Intercept the fetch request to prevent actual generation
        await page.route('**/generate-illustrations', route => {
            route.fulfill({
                status: 200,
                contentType: 'application/json',
                body: JSON.stringify({ status: 'generating', message: 'Test mode' })
            });
        });

        await confirmBtn.click();
        await page.waitForTimeout(200);

        // Check localStorage
        const savedSettings = await page.evaluate(() => {
            return localStorage.getItem('illustration_settings');
        });

        if (!savedSettings) {
            throw new Error('Settings not saved to localStorage');
        }

        const settings = JSON.parse(savedSettings);
        if (settings.numImages !== 5 || settings.style !== 'anime') {
            throw new Error(`Settings not saved correctly: ${JSON.stringify(settings)}`);
        }
        console.log('   ✓ Settings saved to localStorage');
        console.log(`   ✓ Saved values: numImages=${settings.numImages}, style="${settings.style}"\n`);
        testsPassed++;

        // Test 8: Test dialog reopens with saved settings
        console.log('📍 Test 8: Test dialog reopens with saved settings');

        // Reload page to clear state
        await page.reload();
        await page.waitForLoadState('networkidle');

        // Open dialog again
        await page.click('#actions-menu-btn');
        await page.waitForTimeout(200);
        await page.click('#generate-illustrations-btn');
        await page.waitForTimeout(200);

        // Check if fields are populated with saved values
        const reloadedNumValue = await page.inputValue('#num-images');
        const reloadedStyleValue = await page.inputValue('#image-style');

        if (reloadedNumValue !== '5' || reloadedStyleValue !== 'anime') {
            throw new Error(`Settings not loaded from localStorage: num=${reloadedNumValue}, style=${reloadedStyleValue}`);
        }
        console.log('   ✓ Dialog reopened with saved settings');
        console.log(`   ✓ Loaded values: numImages=${reloadedNumValue}, style="${reloadedStyleValue}"\n`);
        testsPassed++;

        // Test 9: Test cancel button
        console.log('📍 Test 9: Test cancel button');
        const cancelBtn = await page.$('button:has-text("Cancel")');
        await cancelBtn.click();
        await page.waitForTimeout(200);

        const isHiddenAfterCancel = await page.$eval('#illustration-dialog',
            el => el.classList.contains('hidden')
        );
        if (!isHiddenAfterCancel) {
            throw new Error('Dialog should be hidden after clicking cancel');
        }
        console.log('   ✓ Dialog closes when clicking cancel\n');
        testsPassed++;

        // Test 10: Test validation
        console.log('📍 Test 10: Test form validation');

        // Open dialog
        await page.click('#actions-menu-btn');
        await page.waitForTimeout(200);
        await page.click('#generate-illustrations-btn');
        await page.waitForTimeout(200);

        // Try to set invalid number
        await page.fill('#num-images', '15'); // Max is 10

        // Click confirm and check for alert
        page.on('dialog', async dialog => {
            console.log(`   ✓ Validation alert shown: "${dialog.message()}"`);
            await dialog.accept();
        });

        await page.click('button:has-text("Generate")');
        await page.waitForTimeout(500);

        // Dialog should still be visible (validation failed)
        const stillVisible = await page.$eval('#illustration-dialog',
            el => !el.classList.contains('hidden')
        );
        if (!stillVisible) {
            throw new Error('Dialog should remain visible after validation failure');
        }
        console.log('   ✓ Form validation prevents invalid values\n');
        testsPassed++;

        // Test 11: Check "Generate All Illustrations" also shows dialog
        console.log('📍 Test 11: Test "Generate All Illustrations" button');

        // Close current dialog
        await page.click('button:has-text("Cancel")');
        await page.waitForTimeout(200);

        // Try to open book menu
        const bookMenuBtn = await page.$('#book-menu-btn');
        if (bookMenuBtn) {
            // Hover to show the button (it has opacity-0 group-hover:opacity-100)
            await page.hover('.group:has(#book-menu-btn)');
            await page.waitForTimeout(200);

            await bookMenuBtn.click();
            await page.waitForTimeout(200);

            const generateAllBtn = await page.$('button:has-text("Generate All Illustrations")');
            if (generateAllBtn) {
                await generateAllBtn.click();
                await page.waitForTimeout(200);

                const dialogVisible = await page.$eval('#illustration-dialog',
                    el => !el.classList.contains('hidden')
                );
                if (!dialogVisible) {
                    throw new Error('Dialog should open for "Generate All Illustrations"');
                }
                console.log('   ✓ Dialog opens for "Generate All Illustrations"\n');
                testsPassed++;
            } else {
                console.log('   ⚠️  "Generate All Illustrations" button not found (may be hidden)\n');
            }
        } else {
            console.log('   ⚠️  Book menu button not found\n');
        }

    } catch (error) {
        console.error(`\n❌ Test failed: ${error.message}\n`);
        testsFailed++;
    } finally {
        await browser.close();
    }

    // Print summary
    console.log('═══════════════════════════════════════════════════');
    console.log('📊 Test Summary:');
    console.log(`   ✓ Passed: ${testsPassed}`);
    console.log(`   ✗ Failed: ${testsFailed}`);
    console.log(`   Total:  ${testsPassed + testsFailed}`);

    if (testsFailed === 0) {
        console.log('\n✅ All tests passed!');
        process.exit(0);
    } else {
        console.log('\n❌ Some tests failed');
        process.exit(1);
    }
}

// Run tests
testCustomizationDialog().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});

File: test_customization.py
"""
Playwright test for image generation customization dialog

Tests:
1. Dialog appears when clicking generate buttons
2. LocalStorage persistence works
3. Form validation works
4. Settings are passed to backend correctly
"""

import asyncio
import sys

try:
    from playwright.async_api import async_playwright
except ImportError:
    print("❌ Playwright not installed. Install with: pip install playwright")
    sys.exit(1)


async def test_customization_dialog():
    print("🧪 Starting Playwright tests for customization dialog...\n")

    tests_passed = 0
    tests_failed = 0

    async with async_playwright() as p:
        try:
            # Launch browser
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context()
            page = await context.new_page()

            # Test 1: Navigate to library page
            print("📍 Test 1: Navigate to library page")
            await page.goto("http://127.0.0.1:8123/")
            await page.wait_for_load_state("networkidle")

            title = await page.title()
            print(f"   ✓ Page loaded: {title}")
            tests_passed += 1

            # Check if there are any books
            book_links = await page.query_selector_all('a[href*="/read/"]')

            if not book_links:
                print("   ⚠️  No books found - skipping reader page tests")
                print("   ℹ️  To fully test, process an EPUB first with: uv run reader3.py <book.epub>\n")

                print("📊 Test Summary:")
                print(f"   ✓ Passed: {tests_passed}")
                print(f"   ✗ Failed: {tests_failed}")
                print(f"   ⚠️  Skipped: Reader page tests (no books available)")

                await browser.close()
                return tests_passed, tests_failed

            print(f"   ✓ Found {len(book_links)} book(s)\n")

            # Test 2: Navigate to first book
            print("📍 Test 2: Navigate to book reader")
            await book_links[0].click()
            await page.wait_for_load_state("networkidle")
            print("   ✓ Reader page loaded\n")
            tests_passed += 1

            # Test 3: Check if customization dialog exists
            print("📍 Test 3: Check customization dialog exists in DOM")
            dialog = await page.query_selector("#illustration-dialog")
            if not dialog:
                raise Exception("Customization dialog not found in DOM")

            is_hidden = await dialog.evaluate("el => el.classList.contains('hidden')")
            if not is_hidden:
                raise Exception("Dialog should be hidden by default")

            print("   ✓ Customization dialog exists and is hidden by default\n")
            tests_passed += 1

            # Test 4: Check dialog fields exist
            print("📍 Test 4: Check dialog form fields")
            num_images_input = await page.query_selector("#num-images")
            style_input = await page.query_selector("#image-style")

            if not num_images_input or not style_input:
                raise Exception("Dialog form fields not found")

            print("   ✓ Number of images input found")
            print("   ✓ Style input found\n")
            tests_passed += 1

            # Test 5: Open dialog via floating action button
            print("📍 Test 5: Open dialog via floating action button")

            action_button = await page.query_selector("#actions-menu-btn")
            if not action_button:
                raise Exception("Floating action button not found")

            await action_button.click()
            await page.wait_for_timeout(300)

            generate_btn = await page.query_selector("#generate-illustrations-btn")
            if not generate_btn:
                raise Exception("Generate illustrations button not found")

            await generate_btn.click()
            await page.wait_for_timeout(300)

            is_visible = await page.evaluate(
                "() => !document.getElementById('illustration-dialog').classList.contains('hidden')"
            )
            if not is_visible:
                raise Exception("Dialog should be visible after clicking generate button")

            print("   ✓ Dialog opens when clicking generate button\n")
            tests_passed += 1

            # Test 6: Test form inputs
            print("📍 Test 6: Test form inputs")
            await page.fill("#num-images", "5")
            await page.fill("#image-style", "anime")

            num_value = await page.input_value("#num-images")
            style_value = await page.input_value("#image-style")

            if num_value != "5" or style_value != "anime":
                raise Exception(f"Form inputs not working: num={num_value}, style={style_value}")

            print("   ✓ Number of images set to 5")
            print("   ✓ Style set to 'anime'\n")
            tests_passed += 1

            # Test 7: Test localStorage persistence
            print("📍 Test 7: Test localStorage persistence")

            # Intercept the fetch request to prevent actual generation
            await page.route("**/generate-illustrations", lambda route: route.fulfill(
                status=200,
                content_type="application/json",
                body='{"status": "generating", "message": "Test mode"}'
            ))

            confirm_btn = await page.query_selector('button:has-text("Generate")')
            await confirm_btn.click()
            await page.wait_for_timeout(300)

            # Check localStorage
            saved_settings = await page.evaluate("() => localStorage.getItem('illustration_settings')")
            if not saved_settings:
                raise Exception("Settings not saved to localStorage")

            import json
            settings = json.loads(saved_settings)
            if settings.get("numImages") != 5 or settings.get("style") != "anime":
                raise Exception(f"Settings not saved correctly: {settings}")

            print("   ✓ Settings saved to localStorage")
            print(f"   ✓ Saved values: numImages={settings['numImages']}, style=\"{settings['style']}\"\n")
            tests_passed += 1

            # Test 8: Test dialog reopens with saved settings
            print("📍 Test 8: Test dialog reopens with saved settings")

            await page.reload()
            await page.wait_for_load_state("networkidle")

            # Open dialog again
            await page.click("#actions-menu-btn")
            await page.wait_for_timeout(300)
            await page.click("#generate-illustrations-btn")
            await page.wait_for_timeout(300)

            reloaded_num_value = await page.input_value("#num-images")
            reloaded_style_value = await page.input_value("#image-style")

            if reloaded_num_value != "5" or reloaded_style_value != "anime":
                raise Exception(f"Settings not loaded: num={reloaded_num_value}, style={reloaded_style_value}")

            print("   ✓ Dialog reopened with saved settings")
            print(f"   ✓ Loaded values: numImages={reloaded_num_value}, style=\"{reloaded_style_value}\"\n")
            tests_passed += 1

            # Test 9: Test cancel button
            print("📍 Test 9: Test cancel button")
            cancel_btn = await page.query_selector('button:has-text("Cancel")')
            await cancel_btn.click()
            await page.wait_for_timeout(300)

            is_hidden_after_cancel = await page.evaluate(
                "() => document.getElementById('illustration-dialog').classList.contains('hidden')"
            )
            if not is_hidden_after_cancel:
                raise Exception("Dialog should be hidden after clicking cancel")

            print("   ✓ Dialog closes when clicking cancel\n")
            tests_passed += 1

            # Test 10: Test validation
            print("📍 Test 10: Test form validation")

            await page.click("#actions-menu-btn")
            await page.wait_for_timeout(300)
            await page.click("#generate-illustrations-btn")
            await page.wait_for_timeout(300)

            # Set invalid number
            await page.fill("#num-images", "15")  # Max is 10

            # Handle alert
            page.on("dialog", lambda dialog: asyncio.create_task(handle_dialog(dialog)))

            async def handle_dialog(dialog):
                print(f"   ✓ Validation alert shown: \"{dialog.message}\"")
                await dialog.accept()

            await page.click('button:has-text("Generate")')
            await page.wait_for_timeout(500)

            # Dialog should still be visible (validation failed)
            still_visible = await page.evaluate(
                "() => !document.getElementById('illustration-dialog').classList.contains('hidden')"
            )
            if not still_visible:
                raise Exception("Dialog should remain visible after validation failure")

            print("   ✓ Form validation prevents invalid values\n")
            tests_passed += 1

            await browser.close()

        except Exception as error:
            print(f"\n❌ Test failed: {error}\n")
            tests_failed += 1
            if 'browser' in locals():
                await browser.close()

    return tests_passed, tests_failed


async def main():
    tests_passed, tests_failed = await test_customization_dialog()

    # Print summary
    print("═══════════════════════════════════════════════════")
    print("📊 Test Summary:")
    print(f"   ✓ Passed: {tests_passed}")
    print(f"   ✗ Failed: {tests_failed}")
    print(f"   Total:  {tests_passed + tests_failed}")

    if tests_failed == 0:
        print("\n✅ All tests passed!")
        sys.exit(0)
    else:
        print("\n❌ Some tests failed")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
